opal/mca/btl/ofi/btl_ofi_providers.h

```diff
@@ -0,0 +1,98 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2018      Intel, Inc, All rights reserved
+ * Copyright (c) 2018      Los Alamos National Security, LLC. All rights
+ *                         reserved.
+ *
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "btl_ofi.h"
+
+#define MCA_BTL_OFI_REQUIRED_CAPS       (FI_RMA | FI_ATOMIC)
+#define MCA_BTL_OFI_REQUESTED_MR_MODE   (FI_MR_UNSPEC)
+
+/* handle provider specific hints */
+static void provider_hints_handler(struct fi_info *hints);
+
+
+/* info validation after fi_getinfo() */
+static int validate_info(struct fi_info *info);
+static int socks_validate_info(struct fi_info *info);
+static int psm2_validate_info(struct fi_info *info);
+static int gni_validate_info(struct fi_info *info);
+
+
+/* functions below */
+
+static int socks_validate_info(struct fi_info *info)
+{
+    /* placeholder */
+    return OPAL_SUCCESS;
+}
+
+static int psm2_validate_info(struct fi_info *info)
+{
+    /* placeholder */
+    return OPAL_SUCCESS;
+}
+
+static int gni_validate_info(struct fi_info *info)
+{
+    /* placeholder */
+    return OPAL_SUCCESS;
+}
+
+/* validate information returned from fi_getinfo().
+ * return OPAL_ERROR if we dont have what we need. */
+static int validate_info(struct fi_info *info)
+{
+    int mr_mode;
+    char *prov_name = info->fabric_attr->prov_name;
+
+    /* we need exactly all the required bits */
+    if ((info->caps & MCA_BTL_OFI_REQUIRED_CAPS) != MCA_BTL_OFI_REQUIRED_CAPS)
+        return OPAL_ERROR;
+
+    /* we need FI_EP_RDM */
+    if (info->ep_attr->type != FI_EP_RDM)
+        return OPAL_ERROR;
+
+    mr_mode = info->domain_attr->mr_mode;
+
+    if (!(mr_mode == FI_MR_BASIC || mr_mode == FI_MR_SCALABLE ||
+         (mr_mode & ~(FI_MR_VIRT_ADDR | FI_MR_ALLOCATED | FI_MR_PROV_KEY)) == 0))
+        return OPAL_ERROR;
+    }
+
+    /* provider specifics */
+    if (!strcmp(prov_name, "sockets")) {
+        return socks_validate_info(info);
+    } else if (!strcmp(prov_name, "psm2")) {
+        return psm2_validate_info(info);
+    } else if (!strcmp(prov_name, "gni")) {
+        return gni_validate_info(info);
+    }
+
+    return OPAL_SUCCESS;
+}
+
+/* fill out specific hints needed for each provider */
+static void provider_hints_handler(struct fi_info *hints)
+{
+    char *prov_name;
+
+    assert(hints->fabric_attr);
+    assert(hints->domain_attr);
+    assert(hints->fabric_attr->prov_name);
+
+    prov_name = hints->fabric_attr->prov_name;
+
+    if (!strcmp(prov_name, "gni")) {
+        hints->domain_attr->mr_mode = FI_MR_BASIC;
```

Comment:
This seems dangerous -- if you *have* to hard-code hints on a per-provider basis, you should probably have MCA vars to override them, too (because they will change over time).

Reply:
We are dealing with multiple providers who has their own setting to satisfy. I don't see how this is different from when we hard-coded endpoint type or caps. 

I can have the MCA to override the MR setting. That won't be easy as it is a flag not an enum but I can try.

Reply:
@thananon let's discuss this next week on our call.

Reply:
@hppritcha I would like to get this done as soon as possible.

Is there any suggestion for me that I can do now? I don't really like `!strcmp` but that's the only way I can see as the provider `prov_name` field is a string.

As for MCA override, I think the easy work around is having one to force FI_MR_BASIC and make it a level 8+. All the provider on 1.5 right now supports FI_MR_BASIC so it should be pretty safe.

Reply:
@hppritcha 

I just realized we dont need to hard code FI_MR_BASIC for gni anymore as now we requested FI_MR_UNSPEC and deal with that in `validate_info()`.  So this can go away.

Fixed in upcoming commit.